### ISSUE TITLE
Replace some subtensors with DimShuffle

### ISSUE THREAD

**Issue by abergeron** [Theano/Theano#4162]
Subtensors like `a[:, :, None, :]` could be done with `a.dimsuffle(0, 1, 'x', 2)` which is faster and better optimized.

See https://github.com/abergeron/Theano/commit/4c41ccddca298b3d5b859906347f82282987372d for an example.

**Comment by lamblin** [comment]
This should be done around [here](https://github.com/Theano/Theano/blob/149d008d528277988867089cb607105e706910c4/theano/tensor/var.py#L506) when calling `__getitem__` in the first place.
For some reason, the existing code may not avoid all subtensors.

**Comment by lamblin** [comment]
So, apparently, it keeps a useless subtensor with all ":":
```python
>>> theano.printing.debugprint(a[:, :, None, :])
Subtensor{::, ::, ::, ::} [id A] ''   
 |DimShuffle{0,1,x,2} [id B] ''   
   |<TensorType(float32, 3D)> [id C]
```

**Comment by abergeron** [comment]
Then we should probably fix the `__getitem__` code to not do that.  It should also handle the case when we index with a 0 on a broadcastable dimension.

**Comment by Sentient07** [comment]
@abergeron , I'd like to take this up if no one is currently working on this. :)

**Comment by lamblin** [comment]
I agree we should fix `__getitem__` to not insert the useless subtensor.
For indexing with 0 on a broadcastable dimension, I would rather use an optimization for that, if only because the variables could be replaced by a `clone` or using `givens`.

**Comment by Sentient07** [comment]
@lamblin I'd like to take this issue. Shall I go ahead ?

**Comment by fvisin** [comment]
@Sentient07 sorry for the late reply! Thank you for volunteering, unfortunately the issue has already being assigned and is currently being worked on.

**Comment by Sentient07** [comment]
@fvisin No problem. :)

**Comment by fvisin** [comment]
This should be a good spot to place the optimization: https://github.com/Theano/Theano/blob/149d008d528277988867089cb607105e706910c4/theano/tensor/opt.py#L1805

The goal is to detect indexing with 0 or -1 on broadcastable dimensions, which boils down to dropping that dimension (so we don't need a subtensor, at least on that axis).

**Comment by SinaHonari** [comment]
Considering the problem with a[:, :, None, :], if I replicate it, it seems like _pattern_ and _view_ are correct and I get [0, 1, 'x', 2] for _pattern_ and DimShuffle{0,1,x,2}.0 for _view_. I'm not sure if _new_args_ is correct, though, I'm getting [slice(None, None, None), slice(None, None, None), slice(None, None, None), slice(None, None, None)] for it. If it's colon, it just passes that in line 523 (new_args.append(arg)). What should be the content of _new_args_ in the correct case?

**Comment by fvisin** [comment]
`[slice(None, None, None), slice(None, None, None), slice(None, None, None), slice(None, None, None)]` is the expected result, because the `:` indexes correspond to `slice(None, None, None)` and `None` is replaced with `slice(None, None, None)` [here](https://github.com/Theano/Theano/blob/149d008d528277988867089cb607105e706910c4/theano/tensor/var.py#L519). 

What you need to do is detect that pattern (i.e. detect when all the args are either `None` or `slice(None, None, None)`) and replace the op as in the example.

**Comment by SinaHonari** [comment]
@fvisin: I made the first commit.

**Comment by fvisin** [comment]
Fixed in #4356
